# Neural Networks from Scratch
**Oliver Mazariegos**

# Neural Networks from Scratch
## Clothing Classification with Fashion-MNIST

This notebook builds a neural network capable of classifying clothing items from images.
The dataset is the [Fashion-MNIST](https://github.com/zalandoresearch/fashion-mnist) collection, sourced from Kaggle and the original Zalando Research repository.
Every algorithm — from forward pass to backpropagation — is implemented manually using NumPy.


## Libraries

In [1]:
import numpy as np
import mnist_reader
from functools import reduce
from scipy import optimize as op
from sklearn.model_selection import train_test_split

## Matrix Utility Functions

In [2]:
flatten_list_of_arrays = lambda list_of_arrays: reduce (
    lambda acc, v: np.array([*acc.flatten(), *v.flatten()],dtype=np.float64),
    list_of_arrays
)

def flatten_matrices(matrices):
    flat = []
    for matrix in matrices:
        flat = [*flat, *(matrix.flatten())]
    return np.array(flat).flatten()
    


def inflate_matrixes(flat_thetas, shapes):
    layers = len(shapes) + 1
    sizes = [shape[0] * shape[1] for shape in shapes]
    steps = np.zeros(layers, dtype=int)
    for i in range(layers - 1):
        steps[i+1] = steps[i] + sizes[i]
    return [
        flat_thetas[steps[i]:steps[i+1]].reshape(*shapes[i])
        for i in range(layers - 1)
    ]

The functions above flatten and inflate weight matrices.
The network uses multiple weight matrices (one per layer transition). The `scipy.optimize.minimize` optimizer expects a single flat 1-D array of parameters, so these helpers convert back and forth between the flat representation and the list of 2-D matrices.


## Activation Function

In [3]:
def sigmoid(x):
    return 1.0/(1 + np.exp(-x))

The **sigmoid** function maps any real value to the range $(0, 1)$ and is used as the activation function for all neurons in this network:

$$\sigma(x) = \frac{1}{1 + e^{-x}}$$


## Feed Forward

**Feed Forward** computes the activation matrix for every layer of the network.
The first layer is the input (the image pixels) and the final layer produces the class prediction.

For each layer $l$:

$$a^{(l+1)} = \sigma\!\left(\begin{bmatrix}1 \mid a^{(l)}\end{bmatrix} \cdot \Theta^{(l)\,T}\right)$$

A bias column of ones is prepended to the activation matrix before the dot product.


In [4]:
def feed_forward(thetas, X):
    # List of activation matrices, starting with the input
    a = [X] 
    for i in range(len(thetas)): 
        a.append(
            sigmoid(
                np.matmul(
                    # Prepend bias column of ones
                    np.hstack((
                        np.ones(len(a[i])).reshape(len(a[i]), 1),
                        a[i]
                    )),
                    thetas[i].T
                )
            )
        )
    return a 

## Backpropagation

**Backpropagation** computes how much each weight contributed to the final error by propagating the error signal backwards through the network.
It produces the gradient matrices $\Delta$, which `scipy.optimize.minimize` uses to update the weights.

For each layer (in reverse):

$$\delta^{(l)} = \left(\delta^{(l+1)} \cdot \Theta^{(l)}_{[:,1:]}\right) \odot a^{(l)} \odot (1 - a^{(l)})$$

The gradient for layer $l$:

$$\frac{\partial J}{\partial \Theta^{(l)}} = \frac{1}{m}\,\delta^{(l+1)\,T} \cdot \begin{bmatrix}1 \mid a^{(l)}\end{bmatrix}$$


In [5]:
def back_propagation(flat_thetas, shapes, X, Y):
    gradient = []
    m, layers = len(X), len(shapes) + 1
    thetas = inflate_matrixes(flat_thetas, shapes)
    a = feed_forward(thetas, X)
    # Initialize error terms; last layer error is prediction minus ground truth
    deltas = [*range(layers - 1), a[-1] - Y]
    # Propagate error backwards from the second-to-last layer to the first hidden layer
    for i in range(layers - 2, 0, -1):
        deltas[i] = np.matmul(deltas[i + 1], thetas[i][:, 1:]) * (a[i] * (1 - a[i]))
    # Compute gradient matrices for each layer
    for i in range(layers - 1):
        gradient.append(
            np.matmul(
                deltas[i + 1].T,
                np.hstack((  # prepend bias column
                    np.ones(len(a[i])).reshape(len(a[i]), 1),
                    a[i]
                ))
            ) / m
        )
    return flatten_matrices(gradient)

## Cost Function

The **cross-entropy cost function** measures how well the network predictions match the true labels.

$$J(\Theta) = -\frac{1}{m} \sum_{i=1}^{m} \left[ y^{(i)} \log(h^{(i)}) + (1 - y^{(i)}) \log(1 - h^{(i)}) \right]$$


In [6]:
def cost_function(flat_thetas, shapes, X, Y):
    a = feed_forward(
        inflate_matrixes(flat_thetas, shapes),
        X
    )
    return -(Y * np.log(a[-1]) + (1 - Y) * np.log(1 - a[-1])).sum() / len(X)

## Dataset

Each image is 28×28 pixels, flattened to a vector of 784 values. Each pixel is an integer in the range [0, 255] (0 = black, 255 = white).
The dataset contains 10 clothing categories:

| Label | Class |
|-------|-------|
| 0 | T-shirt/top |
| 1 | Trouser |
| 2 | Pullover |
| 3 | Dress |
| 4 | Coat |
| 5 | Sandal |
| 6 | Shirt |
| 7 | Sneaker |
| 8 | Bag |
| 9 | Ankle boot |


## Training Set

The training set contains **60,000** samples. Pixel values are divided by 1000 to keep them in a small range, which prevents the sigmoid derivative from saturating to near-zero gradients.


In [7]:
X_train, y_train = mnist_reader.load_mnist('',kind='train')
X = X_train / 1000
m, n = X.shape

The network outputs a vector of shape `(10,)` (one probability per class), but the dataset labels are integers in `[0, 9]`.
The labels are one-hot encoded into a matrix of shape `(m, 10)` so predictions and ground truth can be compared directly.


In [8]:
Y = np.zeros((X.shape[0], 10))
for i in range(m):
    Y[i][y_train[i]] = 1

In [9]:
Y[0]

array([0., 0., 0., 0., 0., 0., 0., 0., 0., 1.])

## Network Architecture

The network has **100 hidden neurons** and **10 output neurons** (one per clothing class).

| Layer | Neurons |
|-------|---------|
| Input | 784 |
| Hidden | 100 |
| Output | 10 |

Weights are initialized randomly with small values (divided by 1000) to break symmetry.


In [10]:
HIDDEN_NEURONS = 100
OUTPUT_NEURONS = 10

theta_shapes = np.array([
    [HIDDEN_NEURONS, n + 1],
    [OUTPUT_NEURONS, HIDDEN_NEURONS + 1]
])


flat_thetas = flatten_list_of_arrays(
    [np.random.rand(*theta_shape) / 1000 for theta_shape in theta_shapes]
)

## Model Optimization

`scipy.optimize.minimize` is used to optimize the weights. It receives:

- The cost function to minimize
- An initial flat array of randomly initialized weights
- The training data `X` and labels `Y`
- The gradient function (backpropagation)
- The optimization method: **L-BFGS-B** (a quasi-Newton method well-suited for large-scale problems)

Training runs for up to **400 iterations**.


In [11]:
result = op.minimize(
    fun=cost_function,
    x0=flat_thetas,
    args=(theta_shapes,X,Y),
    method='L-BFGS-B',
    jac=back_propagation,
    options={'disp':True, 'maxiter':400}
)

/tmp/ipykernel_12284/1191445197.py:1: DeprecationWarning: scipy.optimize: The `disp` and `iprint` options of the L-BFGS-B solver are deprecated and will be removed in SciPy 1.18.0.
  result = op.minimize(


## Trained Model

The optimizer returns the optimal weights as a flat array. These are inflated back into the original matrix shapes for use in predictions.


In [12]:
result.x

array([-1.58561830e+00,  9.95003080e-04, -3.82249788e-05, ...,
       -6.75345522e+00,  1.00168121e+00,  1.39204341e+00], shape=(79510,))

In [13]:
modelo = inflate_matrixes(result.x,theta_shapes)

## Test Set

The test set contains **10,000** samples not seen during training. It has the same structure as the training set.


In [14]:
X_test, y_test = mnist_reader.load_mnist('', kind='t10k')
X = X_test / 1000
m, n = X.shape
Y = np.zeros((X.shape[0], 10))
for i in range(m):
    Y[i][y_test[i]] = 1

## Predictions

In [15]:
prediccion = feed_forward(modelo,X)[-1]

## Model Accuracy

Accuracy is the fraction of correctly classified samples:

$$\text{Accuracy} = \frac{\text{Correct predictions}}{\text{Total samples}}$$


In [16]:
def accuracy(predictions, labels):
    correct = 0
    for i in range(len(labels)):
        if np.argmax(labels[i]) == np.argmax(predictions[i]):
            correct += 1
    return correct / len(labels)
        

In [17]:
print(f"Model accuracy on test set: {accuracy(prediccion, Y):.4f}")

Model accuracy on test set: 0.8792


## Conclusion

A neural network was successfully trained from scratch to classify Fashion-MNIST clothing images. With only 400 optimization iterations, 100 hidden neurons, and a manually implemented backpropagation + feed-forward pipeline, the model achieves approximately **88% accuracy** on the test set.
